### PR TITLE
[PT 8] Admin flow and tests

### PR DESCRIPTION
**Pull request by KauhareniaNastassia**
Add users page for admin
Add changing user role for super admin
Add orders page for admin
Add changing order status
Add menu page for admin
Add feature for adding and deleting menu items
Add page for editing menu items
Add unit, widget and integration tests

https://github.com/KauhareniaNastassia/food_delivery_app/assets/104850152/41a806de-b12b-4d79-a3a9-f98f9afcd3e3
https://github.com/KauhareniaNastassia/food_delivery_app/assets/104850152/35b55c4c-17a6-453d-a974-18af941d59f1

### CODE REVIEW

File: core/lib/constants/app_constants.dart [hunk 1]
```diff
@@ -16,4 +16,17 @@ class AppConstants {
     'Lunch',
     'For a company',
   ];
+
```
Comment (SashaPilevich) [inline core/lib/constants/app_constants.dart:11]: fetch this info from Firebase

File: core/lib/validation/validation.dart [hunk 2]
```diff
@@ -34,3 +34,43 @@ String? nameValidation({
   }
   return null;
 }
+
+String? menuItemTitleValidation(String? title) {
+  if (title!.isEmpty) {
```
Comment (SashaPilevich) [inline core/lib/validation/validation.dart]: change ! on ??

File: core_ui/lib/src/widgets/bottom_bar.dart [hunk 3]
```diff
@@ -28,26 +31,28 @@ class OrderBottomBar extends StatelessWidget {
...
-                      ),
-                    ],
-                  ),
+                  child: totalPrice == null
+                      ? const SizedBox()
+                      : Row(
+                          mainAxisAlignment: MainAxisAlignment.spaceBetween,
+                          children: <Widget>[
+                            Text(
+                              title!,
```
Comment (SashaPilevich) [inline core_ui/lib/src/widgets/bottom_bar.dart]: change ! on ??

File: core_ui/lib/src/widgets/custom_bottom_navigation_bar.dart [hunk 4]
```diff
@@ -4,12 +4,14 @@ import 'package:flutter/material.dart';
 
 class CustomBottomNavigationBar extends StatefulWidget {
   final int currentIndex;
+  final bool isCustomer;
```
Comment (SashaPilevich) [inline core_ui/lib/src/widgets/custom_bottom_navigation_bar.dart:7]: convert to stateless widget

File: core_ui/lib/src/widgets/custom_drop_down_field.dart [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      icon: Icon(
+        Icons.arrow_drop_down_rounded,
+        size: mediaQueryData.size.height * 0.05,
+      ),
+      style: theme.textTheme.displayMedium,
+      underline: Container(
+        height: mediaQueryData.size.height * 0.002,
+        color: theme.primaryColor,
+      ),
+      onChanged: (String? value) => onChanged(value!),
```
Comment (SashaPilevich) [inline core_ui/lib/src/widgets/custom_drop_down_field.dart]: use ??

File: core_ui/lib/src/widgets/custom_drop_down_field.dart [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      items: listOfItems.map<DropdownMenuItem<String>>(
```
Comment (SashaPilevich) [inline core_ui/lib/src/widgets/custom_drop_down_field.dart]: create a separate variable for this action

File: data/lib/entity/order_item_entity/order_item_for_admin.dart [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+  factory OrderItemForAdminEntity.fromJson(Map<String, dynamic> json) {
+    return OrderItemForAdminEntity(
+      userId: json['userId'],
+      email: json['email'],
+      orderItem: OrderItemEntity.fromJson(json['orderItem']),
+    );
+  }
+
+  Map<String, dynamic> toMap() {
```
Comment (SashaPilevich) [inline data/lib/entity/order_item_entity/order_item_for_admin.dart]: rename toJson

File: data/lib/repositories/admin_panel_repository_impl/admin_panel_repository_impl.dart [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    required FirebaseFireStoreProvider firebaseFireStoreProvider,
+  }) : _firebaseFireStoreProvider = firebaseFireStoreProvider;
+
+  @override
+  Future<List<UserInfoModel>> fetchUsers() async {
+    final List<UserInfoEntity> usersFromDB =
+        await _firebaseFireStoreProvider.fetchUsers();
+    final List<UserInfoModel> users = usersFromDB
+        .map(
+          (UserInfoEntity e) => UserInfoMapper.toModel(e),
```
Comment (SashaPilevich) [inline data/lib/repositories/admin_panel_repository_impl/admin_panel_repository_impl.dart:19]: rename

File: features/admin_panel/lib/src/bloc/event.dart [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+class InitUsersEvent extends AdminPanelEvent {
+  InitUsersEvent();
+}
+
+class FilterUsersByRoleEvent extends AdminPanelEvent {
+  final String filterValue;
+
+  FilterUsersByRoleEvent({
```
Comment (SashaPilevich) [inline features/admin_panel/lib/src/bloc/event.dart]: const

File: features/admin_panel/lib/src/bloc/state.dart [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        selectedFilter = AppConstants.allUsers,
+        userOrderHistory = [],
+        completedOrdersList = [],
+        unCompletedOrdersList = [],
+        selectedOrdersFilter = AppConstants.orderStatus[0],
+        uploadedMenuItemImage = '',
+        isItemEditing = false,
+        isDataProcessing = false,
+        exception = '',
+        menuItem = MenuItemModel(
```
Comment (SashaPilevich) [inline features/admin_panel/lib/src/bloc/state.dart]: create   .empty()

File: features/admin_panel/lib/src/ui/components/widgets/menu_item_details_for_admin/menu_item_details_content_for_admin.dart [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+                    listOfIngredientControllers: listOfIngredientControllers,
+                    addIngredientController: addIngredientController,
+                    removeIngredientController: removeIngredientController,
+                  ),
+                ],
+              ),
+      ),
+    );
+  }
+}
```
Comment (SashaPilevich) [inline features/admin_panel/lib/src/ui/components/widgets/menu_item_details_for_admin/menu_item_details_content_for_admin.dart:97]: use BlocBuilder

File: features/admin_panel/lib/src/ui/components/widgets/menu_item_details_for_admin/menu_item_details_ingredients_for_admin.dart [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+            : const SizedBox(),
+        ingredients.isNotEmpty
+            ? ListView.builder(
+                shrinkWrap: true,
+                physics: const NeverScrollableScrollPhysics(),
+                itemCount: isEditMode
+                    ? listOfIngredientControllers.length
+                    : ingredients.length,
+                padding: const EdgeInsets.all(0),
+                itemBuilder: (_, index) {
```
Comment (SashaPilevich) [inline features/admin_panel/lib/src/ui/components/widgets/menu_item_details_for_admin/menu_item_details_ingredients_for_admin.dart]: type

File: features/admin_panel/lib/src/ui/components/widgets/menu_item_details_for_admin/menu_item_details_screen_for_admin.dart [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+    setState(() {
+      listOfIngredientControllers[index].clear();
+      listOfIngredientControllers[index].dispose();
+      listOfIngredientControllers.removeAt(index);
+    });
+  }
+
+  void onChangeCategory(String? value) {
+    setState(() {
+      selectedCategory = value!;
```
Comment (SashaPilevich) [inline features/admin_panel/lib/src/ui/components/widgets/menu_item_details_for_admin/menu_item_details_screen_for_admin.dart]: ??

File: features/admin_panel/lib/src/ui/components/widgets/menu_item_details_for_admin/menu_item_details_screen_for_admin.dart [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+
+    return BlocBuilder<AdminPanelBloc, AdminPanelState>(
+      builder: (BuildContext context, AdminPanelState state) {
+        return Form(
+          key: _formKey,
+          child: Scaffold(
+            bottomNavigationBar: MenuItemDetailsBottomBarForAdmin(
+              isEditingMode: adminPanelBloc.state.isItemEditing,
+              onPressCancel: () {
+                if (state.menuItem.id == '') {
```
Comment (SashaPilevich) [inline features/admin_panel/lib/src/ui/components/widgets/menu_item_details_for_admin/menu_item_details_screen_for_admin.dart]: isEmpty

File: features/admin_panel/lib/src/ui/components/widgets/users_page_widgets/user_item_title.dart [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+                ? const LoadingIndicator()
+                : authBloc.state.userRole == AppConstants.userRoles[2]
+                    ? CustomDropDownField(
+                        listOfItems: AppConstants.userRoles,
+                        value: userInfoItem.userRole,
+                        onChanged: (String? value) {
+                          adminPanelBloc.add(
+                            ChangeUserRoleEvent(
+                              userId: userInfoItem.userId,
+                              newUserRoleValue: value!,
```
Comment (SashaPilevich) [inline features/admin_panel/lib/src/ui/components/widgets/users_page_widgets/user_item_title.dart]: ??

File: features/settings/lib/src/ui/user_info/user_name_block.dart [hunk 16]
```diff
@@ -23,12 +25,39 @@ class UserNameBlock extends StatelessWidget {
       crossAxisAlignment: CrossAxisAlignment.start,
       children: <Widget>[
         Text(
-          userName,
+          settingsBloc.state.fontSize > 1 && userName.length > 19
+              ? '${userName.substring(0, 16)}...'
+              : settingsBloc.state.fontSize <= 1 && userName.length > 25
+                  ? '${userName.substring(0, 23)}...'
+                  : userName,
           style: theme.textTheme.titleLarge,
```
Comment (SashaPilevich) [inline features/settings/lib/src/ui/user_info/user_name_block.dart]: refactor this

File: features/settings/lib/src/ui/user_info/user_name_block.dart [hunk 17]
```diff
@@ -23,12 +25,39 @@ class UserNameBlock extends StatelessWidget {
...
+                ],
+              ),
         Text(
-          email,
+          settingsBloc.state.fontSize > 1 && email.length > 21
+              ? '${email.substring(0, 19)}...'
+              : settingsBloc.state.fontSize <= 1 && email.length > 25
+                  ? '${email.substring(0, 22)}...'
+                  : email,
           style: settingsBloc.state.fontSize > 1
```
Comment (SashaPilevich) [inline features/settings/lib/src/ui/user_info/user_name_block.dart]: same

File: features/admin_panel/lib/src/ui/components/widgets/menu_item_details_for_admin/menu_item_details_screen_for_admin.dart (hunk 13 above)
Comment (KauhareniaNastassia) [reply features/admin_panel/lib/src/ui/components/widgets/menu_item_details_for_admin/menu_item_details_screen_for_admin.dart]: bloc

File: core/lib/validation/validation.dart [hunk 18]
```diff
@@ -34,3 +34,43 @@ String? nameValidation({
...
+    return 'numbersAfterDot'.tr();
+  }
+  return null;
+}
+
+String? menuItemDescriptionValidation(String? description) {
+  if (description!.isEmpty) {
+    return 'descriptionIsRequired'.tr();
+  }
+  if (description.length > 500) {
```
Comment (KauhareniaNastassia) [inline core/lib/validation/validation.dart]: const

File: data/lib/providers/firebase_firestore_provider.dart [hunk 19]
```diff
@@ -75,6 +77,147 @@ class FirebaseFireStoreProvider {
       'userId': userInfoEntity.userId,
       'userName': userInfoEntity.userName,
       'email': userInfoEntity.email,
+      'userRole': userInfoEntity.userRole,
+    });
+  }
+
+  ///admin panel
+  Future<List<UserInfoEntity>> fetchUsers() async {
+    final QuerySnapshot<Map<String, dynamic>> userInfoQuerySnapshot =
```
Comment (KauhareniaNastassia) [inline data/lib/providers/firebase_firestore_provider.dart]: naming

File: features/admin_panel/lib/src/ui/components/widgets/menu_item_details_for_admin/menu_item_details_content_for_admin.dart [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+                  SizedBox(height: mediaQueryData.size.height * 0.03),
+                  MenuItemDetailsDescriptionForAdmin(
+                    descriptionController: descriptionController,
+                    description: descriptionController.text,
+                    isEditMode: adminPanelBloc.state.isItemEditing,
+                  ),
+                  SizedBox(height: mediaQueryData.size.height * 0.03),
+                  MenuItemDetailsIngredientsForAdmin(
+                    ingredients: listOfIngredientControllers
+                        .map((TextEditingController ingredientController) =>
```
Comment (KauhareniaNastassia) [inline features/admin_panel/lib/src/ui/components/widgets/menu_item_details_for_admin/menu_item_details_content_for_admin.dart]: formatting

File: features/admin_panel/lib/src/ui/components/widgets/menu_item_details_for_admin/menu_item_details_description_for_admin.dart [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+    final ThemeData theme = Theme.of(context);
+
+    return Column(
+      children: <Widget>[
+        isEditMode
+            ? MenuItemTextField(
+                label: 'description'.tr(),
+                textEditingController: descriptionController,
+                width: mediaQueryData.size.width * 0.9,
+                maxLines: null,
```
Comment (KauhareniaNastassia) [inline features/admin_panel/lib/src/ui/components/widgets/menu_item_details_for_admin/menu_item_details_description_for_admin.dart:29]: ???

**Pull request merged**